Python/Product/AnacondaInstallLauncher/Program.cs

```diff
@@ -0,0 +1,192 @@
+            Console.Error.WriteLine();
+            return ERROR_INVALID_PARAMETER;
+        }
+
+        static int Main(string[] args) {
+            Program p;
+
+            try {
+                p = new Program(args[0], args[1], args[2]);
+#if DEBUG
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex);
+                PrintUsage();
+                return ex.HResult;
+            }
+#else
+            } catch {
+                return PrintUsage();
+            }
+#endif
+
+            Console.CancelKeyPress += (_, e) => {
+                Console.Error.WriteLine("Cancelling wait");
+                p.Cancel();
+            };
+
+            return p.Go();
+        }
+
+        private readonly bool _install;
+        private readonly string _installer, _targetDir;
+        private DateTime _stopAt;
+        private bool _cancel;
+
+        private Program(string command, string installer, string targetDir) {
+            _targetDir = Path.GetFullPath(targetDir);
+            _installer = Path.GetFullPath(installer);
+            if (!File.Exists(_installer)) {
+                Console.Error.WriteLine($"Did not find {_installer}");
+                throw new FileNotFoundException(_installer);
+            }
+            _stopAt = DateTime.UtcNow.AddMinutes(20);
+            _install = command.Equals("install", StringComparison.OrdinalIgnoreCase);
+        }
+
+        int Go() {
+            try {
+                if (_install) {
+                    return DoInstall();
+                } else {
+                    return DoUninstall();
+                }
+            } catch (Exception ex) {
+                Console.Error.WriteLine(ex);
+                Console.Error.WriteLine(ex.ToString());
+                return ex.HResult;
+            }
+        }
+
+        void Cancel() => _cancel = true;
+
+        Process Run(string filename, string arguments) {
+            var psi = new ProcessStartInfo(filename, arguments) {
+                CreateNoWindow = true,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false
+            };
+            Console.Error.WriteLine("Starting:{0}  Filename: {1}{0}  Argument: {2}", Environment.NewLine, psi.FileName, psi.Arguments);
+            var p = Process.Start(psi);
+
+            p.OutputDataReceived += (_, e) => { if (e.Data != null) Console.Write(e.Data); };
+            p.ErrorDataReceived += (_, e) => { if (e.Data != null) Console.Error.Write(e.Data); };
+
+            return p;
+        }
+
+        int DoInstall() {
+            WaitForUninstall(false);
+            if (_cancel) {
+                return ERROR_INSTALL_CANCEL;
+            }
+
+            var p = Run(_installer, string.Format(CultureInfo.InvariantCulture, "/InstallationType=AllUsers /RegisterPython=0 /S /D={0}", _targetDir));
```

Comment:
_targetDir probably needs to be quoted?

Reply:
No. Anaconda installer is weird here - it must be last on the command line and it must be unquoted.

Reply:
Please put that as a comment